drivers/QSPI.cpp

```
@@ -0,0 +1,297 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "drivers/QSPI.h"
+#include "platform/mbed_critical.h"
+
+#if DEVICE_QSPI
+#define IS_BUS_WIDTH_VALID(width)          ((width == 1) || (width == 2) || (width == 4))    
+#define IS_SIZE_VALID(size)                ((size == 8) || (size == 16) || (size == 24) || (size == 32))      
+#define IS_ALT_SIZE_VALID(alt_size)        ((alt_size == 0) || (alt_size == 8) || (alt_size == 16) || (alt_size == 24) || (alt_size == 32))      
+
+namespace mbed {
+
+QSPI* QSPI::_owner = NULL;
+SingletonPtr<PlatformMutex> QSPI::_mutex;  
+  
+QSPI::QSPI(PinName io0, PinName io1, PinName io2, PinName io3, PinName sclk, PinName ssel) :
+        _qspi(),
+        _inst_width(QSPI_DEFAULT_INST_WIDTH),
+        _address_width(QSPI_DEFAULT_ADDRESS_WIDTH),
+        _address_size(QSPI_DEFAULT_ADDRESS_SIZE),
+        _alt_width(QSPI_DEFAULT_ALT_WIDTH),
+        _alt_size(QSPI_DEFAULT_ALT_SIZE),
+        _data_width(QSPI_DEFAULT_DATA_WIDTH),
+        _num_dummy_cycles(QSPI_DEFAULT_DUMMY_CYCLES),
+        _hz(QSPI_DEFAULT_HZ) {
+    // No lock needed in the constructor
+    _qspi_io0 = io0;
+    _qspi_io1 = io1;
+    _qspi_io2 = io2;
+    _qspi_io3 = io3;
+    _qspi_clk = sclk;
+    _qspi_cs = ssel;            
+}
+        
+bool QSPI::configure_format(int inst_width, 
+                     int address_width, int address_size,  
+                     int alt_width, int alt_size,
+                     int data_width,
+                     int dummy_cycles,
+                     int mode ) {
+    if(!IS_BUS_WIDTH_VALID(inst_width)) return false;
+    if(!IS_BUS_WIDTH_VALID(address_width)) return false;
+    if(!IS_SIZE_VALID(address_size)) return false;
+    if(!IS_BUS_WIDTH_VALID(alt_width)) return false;
+    if(!IS_ALT_SIZE_VALID(alt_size)) return false;
+    if(!IS_BUS_WIDTH_VALID(data_width)) return false;
```

is this runtime error or a user error ? Shall we assert instead? I like not asserting as this can be tested easily, but again with only false we are what is wrong, run debug to get details. If we use status qspi type, we get invalid parameter at least

---

This is runtime error. Assert call can be in the APP. We can return a specific error if its helpful, let me look into that.